src/templates/next/layouts/Content/Content.tsx

```diff
@@ -1,33 +1,140 @@
-import { ContentPageSchema } from "~/engine"
+import { ContentPageSchema, IsomerSitemap } from "~/engine"
 import ContentPageHeader from "../../components/shared/ContentPageHeader"
 import Siderail from "../../components/shared/Siderail"
 import TableOfContents from "../../components/shared/TableOfContents"
 import { Skeleton } from "../Skeleton"
 import { renderComponent } from "../render"
+import { BreadcrumbProps, SiderailProps } from "~/common"
+
+const getBreadcrumbFromSiteMap = (
+  sitemap: IsomerSitemap,
+  permalink: string[],
+): BreadcrumbProps => {
+  const breadcrumb = []
+  let node = sitemap
+  let currentPath = ""
+  for (const pathSegment of permalink) {
+    currentPath += "/" + pathSegment
+    const nextNode = node.children?.find(
+      (node) => node.permalink === currentPath,
+    )
+    if (!nextNode) {
+      // TODO: handle this unexpected case where cannot traverse to permalink in the sitemap
+      break
+    }
+    node = nextNode
+    breadcrumb.push({
+      title: node.title,
+      url: node.permalink,
+    })
+  }
+  return { links: breadcrumb }
+}
+
+const getSiderailFromSiteMap = (
+  sitemap: IsomerSitemap,
+  permalink: string[],
+): SiderailProps | null => {
+  let node = sitemap
+  let currentPath = ""
+
+  let i = 0
+  while (i < permalink.length - 1) {
+    currentPath += "/" + permalink[i]
+    const nextNode = node.children?.find(
+      (node) => node.permalink === currentPath,
+    )
+    if (!nextNode) {
+      // TODO: handle this unexpected case where cannot traverse to parent in the sitemap
+      return null
+    }
+    node = nextNode
+    i++
+  }
+  if (!node.children) {
+    // TODO: handle this unexpected case where parent does not contain current page
+    return null
+  }
+  const parentTitle = node.title
+  const parentUrl = node.permalink
+
+  const pages = []
+  // get all siblings of page
+  const pagePath = "/" + permalink.join("/")
+  for (const sibling of node.children) {
+    if (sibling.permalink === pagePath) {
+      pages.push({
+        title: sibling.title,
+        url: sibling.permalink,
+        isCurrent: true,
+        childPages: sibling.children?.map((child) => ({
+          url: child.permalink,
+          title: child.title,
+        })),
+      })
+    } else {
+      pages.push({
+        title: sibling.title,
+        url: sibling.permalink,
+      })
+    }
+  }
+  return {
+    parentTitle,
+    parentUrl,
+    pages,
+  }
+}
+
+const getTableOfContentsFromContent = ({
+  content,
+}: Pick<ContentPageSchema, "content">) => {
+  const items = []
+  for (const block of content) {
+    if (block.type === "heading" && block.level === 2) {
+      items.push({
+        content: block.content,
+        anchorLink: "#" + block.id,
+      })
+    }
+  }
```

Comment:
super nit: maybe can use `content.filter` and `.map` instead

Reply:
oh I originally did that, but couldn't get ts to play nice with the discriminated union, i.e after the filter, the block in map does not know it has `.content` and `.id`